databroker/core.py

```diff
@@ -0,0 +1,558 @@
+from __future__ import print_function
+import six  # noqa
+from collections import defaultdict
+from itertools import chain
+import pandas as pd
+import tzlocal
+import doct as doc
+from pims import FramesSequence, Frame
+from metadatastore.commands import NoRunStop, NoEventDescriptors
+from filestore import handlers, HandlerBase
+import logging
+import numbers
+import boltons.cacheutils
+
+# Toolz and CyToolz have identical APIs -- same test suite, docstrings.
+try:
+    from cytoolz.dicttoolz import merge
+except ImportError:
+    from toolz.dicttoolz import merge
+
+
+logger = logging.getLogger(__name__)
+
+
+def fill_event(fs, event, handler_registry=None, handler_overrides=None):
+    """
+    Populate events with externally stored data.
+
+    Parameters
+    ----------
+    fs : FileStoreRO
+    event : document
+    handler_registry : dict, optional
+        mapping spec names (strings) to handlers (callable classes)
+    handler_overrides : dict, optional
+        mapping data keys (strings) to handlers (callable classes)
+
+    .. warning
+
+       This mutates the event's ``data`` field in-place
+    """
+    if handler_overrides is None:
+        handler_overrides = {}
+    is_external = _external_keys(event.descriptor)
+    mock_registries = {data_key: defaultdict(lambda: handler)
+                       for data_key, handler in handler_overrides.items()}
+    for data_key, value in six.iteritems(event.data):
+        if is_external.get(data_key) is not None:
+            if data_key not in handler_overrides:
```

Comment:
@klauer I think you comment on this codeblock (or similar) but you comment is not showing up on github anymore. The code here is inefficient and problematic, but it's not worse than current master. I'd like to revisit in the next day or two in a separate PR.